Mathlib/LinearAlgebra/ExteriorAlgebra/Basic.lean

```diff
@@ -330,10 +330,123 @@ theorem ιMulti_succ_curryLeft {n : ℕ} (m : M) :
+  refine le_antisymm (by rw [Submodule.span_le]; apply le_trans (ιMulti_range R n) (le_refl _)) ?_
+  rw [Submodule.pow_eq_span_pow_set, Submodule.span_le]
+  refine fun u hu ↦ Submodule.subset_span ?_
+  obtain ⟨f, hfu⟩ := Set.mem_pow.mp hu
+  rw [Set.mem_range, ← hfu]
+  existsi fun i => ιInv (f i).1
+  rw [ιMulti_apply]
+  congr
+  ext i
+  obtain ⟨v, hv⟩ := LinearMap.mem_range.mp (by simp only [SetLike.coe_mem] :
+    (f i).1 ∈ LinearMap.range (ι R (M := M)))
+  rw [← hv, ι_inj]; erw [ι_leftInverse]
+
+/-- Given a linearly ordered family `v` of vectors of `M` and a natural number `n`, produce the
+family of `n`fold exterior products of elements of `v`, seen as members of the exterior algebra. -/
+def ιMulti_family (n : ℕ) {I : Type*} [LinearOrder I] (v : I → M) :
+    {s : Finset I // Finset.card s = n} → ExteriorAlgebra R M :=
+  fun s ↦ ιMulti R n (fun i => v (Finset.orderIsoOfFin _ s.prop i))
+
+variable {R}
+
 /-- An `ExteriorAlgebra` over a nontrivial ring is nontrivial. -/
 instance [Nontrivial R] : Nontrivial (ExteriorAlgebra R M) :=
   (algebraMap_leftInverse M).injective.nontrivial
 
+/-! Functoriality of the exterior algebra. -/
+
+variable {N : Type u4} {N' : Type u5} [AddCommGroup N] [Module R N] [AddCommGroup N'] [Module R N']
+
+/-- The morphism of exterior algebras induced by a linear map. -/
+def map (f : M →ₗ[R] N) : ExteriorAlgebra R M →ₐ[R] ExteriorAlgebra R N :=
+  CliffordAlgebra.map { f with map_app' := fun _ => rfl }
+
+@[simp]
+theorem map_comp_ι (f : M →ₗ[R] N) : (map f).toLinearMap ∘ₗ ι R = ι R ∘ₗ f :=
+  CliffordAlgebra.map_comp_ι _
+
+@[simp]
+theorem map_apply_ι (f : M →ₗ[R] N) (m : M) : map f (ι R m) = ι R (f m) :=
+  CliffordAlgebra.map_apply_ι _ m
+
+@[simp]
+theorem map_apply_ιMulti {n : ℕ} (f : M →ₗ[R] N) (m : Fin n → M) :
+    map f (ιMulti R n m) = ιMulti R n (f ∘ m) := by
+  rw [ιMulti_apply, ιMulti_apply, map_list_prod]
+  simp only [List.map_ofFn, Function.comp, map_apply_ι]
+
+@[simp]
+theorem map_comp_ιMulti {n : ℕ} (f : M →ₗ[R] N) :
+    (map f).toLinearMap.compAlternatingMap (ιMulti R n (M := M)) =
+    (ιMulti R n (M := N)).compLinearMap f := by
+  ext m
+  exact map_apply_ιMulti _ _
+
+@[simp]
+theorem map_id :
+    map LinearMap.id = AlgHom.id R (ExteriorAlgebra R M) :=
+  CliffordAlgebra.map_id 0
+
+@[simp]
+theorem map_comp_map (f : M →ₗ[R] N) (g : N →ₗ[R] N') :
+    AlgHom.comp (map g) (map f) = map (LinearMap.comp g f) :=
+  CliffordAlgebra.map_comp_map _ _
+
+@[simp]
+theorem ι_range_map_map (f : M →ₗ[R] N) :
+    Submodule.map (AlgHom.toLinearMap (map f)) (LinearMap.range (ι R (M := M))) =
+    Submodule.map (ι R) (LinearMap.range f) :=
+  CliffordAlgebra.ι_range_map_map _
+
+open Function in
+/-- For a linear map `f` from `M` to `N`,
+`ExteriorAlgebra.map g` is a retraction of `ExteriorAlgebra.map f` iff
+`g` is a retraction of `f`. -/
+@[simp]
+lemma leftInverse_map_iff {f : M →ₗ[R] N} {g : N →ₗ[R] M} :
+    LeftInverse (map g) (map f) ↔ LeftInverse g f := by
```

Comment:
Can you prove this for `CliffordAlgebra` too?

Reply:
Oh, it's not true (in general)!

Reply:
One direction is very easy (and has the same proof as for the exterior algebra). The other direction is true if `[CliffordAlgebra.ι](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/CliffordAlgebra/Basic.html#CliffordAlgebra.%CE%B9)` is injective, which I thought was true but I'm not used to Clifford algebras (especially over general rings), so who knows. 
When I first wrote the injectivity lemma, I decided against having a Clifford algebra version because my goal was "over a field, `map f` is injective if `f` is injective", which is false in general. But it would be quite easy to add at least one direction of the lemma above for Clifford algebras, if you think that has value (I am not sure myself).

Reply:
> The other direction is true if [CliffordAlgebra.ι](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/CliffordAlgebra/Basic.html#CliffordAlgebra.%CE%B9) is injective, which I thought was true but I'm not used to Clifford algebras

This is true whenever `Q` is induced by a bilinear form, though mathlib only actually states it (via `equivExterior`) for rings with an invertible 2. I have plans to tidy that up, as the proof would all work in the more general case.

I don't know if it's true if `Q` is not induced by a bilinear form.

Reply:
> I think it's worth having one direction of the result, simply because it shortens the proof here.

Ok, just a second.

Reply:
Done, thought probably not optimal.
As for the question of generalizing to Clifford algebras, I think it might be too ambitious for this PR.

Reply:
Okay, it wasn't that bad. I left `mapTriv` and `snd_comp_mapTriv` in the same file for now, but they should probably be moved [there](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/TrivSqZeroExt.html), with better names.